sklearn/cluster/_hdbscan/hdbscan.py

```diff
@@ -0,0 +1,718 @@
+    **metric_params,
+):
+    if metric == "precomputed":
+        # Treating this case explicitly, instead of letting
+        # sklearn.metrics.pairwise_distances handle it,
+        # enables the usage of numpy.inf in the distance
+        # matrix to indicate missing distance information.
+        distance_matrix = X
+    else:
+        distance_matrix = pairwise_distances(
+            X, metric=metric, n_jobs=n_jobs, **metric_params
+        )
+
+    if issparse(distance_matrix):
+        return _hdbscan_sparse_distance_matrix(
+            distance_matrix,
+            min_samples,
+            alpha,
+            **metric_params,
+        )
+
+    mutual_reachability_ = mutual_reachability(distance_matrix, min_samples, alpha)
+
+    min_spanning_tree = mst_linkage_core(mutual_reachability_)
+
+    # Warn if the MST couldn't be constructed around the missing distances
+    if np.isinf(min_spanning_tree.T[2]).any():
+        warn(
+            "The minimum spanning tree contains edge weights with value "
+            "infinity. Potentially, you are missing too many distances "
+            "in the initial distance matrix for the given neighborhood "
+            "size.",
+            UserWarning,
+        )
+
+    return _process_mst(min_spanning_tree)
+
+
+def _hdbscan_sparse_distance_matrix(
+    X,
+    min_samples=5,
+    alpha=1.0,
+    **metric_params,
+):
+    # Compute sparse mutual reachability graph
+    # if max_dist > 0, max distance to use when the reachability is infinite
+    max_dist = metric_params.get("max_dist", 0.0)
+    mutual_reachability_ = sparse_mutual_reachability(
+        X.tolil(), min_points=min_samples, max_dist=max_dist, alpha=alpha
+    )
+    # Check connected component on mutual reachability
+    # If more than one component, it means that even if the distance matrix X
+    # has one component, there exists with less than `min_samples` neighbors
+    if (
+        csgraph.connected_components(
+            mutual_reachability_, directed=False, return_labels=False
+        )
+        > 1
+    ):
+        raise ValueError(
+            "There exists points with fewer than %s neighbors. "
+            "Ensure your distance matrix has non-zero values for "
+            "at least `min_sample`=%s neighbors for each points (i.e. K-nn graph), "
+            "or specify a `max_dist` in `metric_params` to use when distances "
+            "are missing." % (min_samples, min_samples)
+        )
+
+    # Compute the minimum spanning tree for the sparse graph
+    sparse_min_spanning_tree = csgraph.minimum_spanning_tree(mutual_reachability_)
+
+    # Convert the graph to scipy cluster array format
+    nonzeros = sparse_min_spanning_tree.nonzero()
+    nonzero_vals = sparse_min_spanning_tree[nonzeros]
+    min_spanning_tree = np.vstack(nonzeros + (nonzero_vals,)).T
+
+    # Sort edges of the min_spanning_tree by weight
+    min_spanning_tree = min_spanning_tree[np.argsort(min_spanning_tree.T[2]), :][0]
```

Comment:
I had to debug to understand what was going on here. I would propose to refactor as:

```suggestion
    edges_sorted_indices = np.argsort(sparse_min_spanning_tree.data)
    rows, cols = sparse_min_spanning_tree.nonzero()
    min_spanning_tree = np.vstack(
        (
            rows[edges_sorted_indices],
            cols[edges_sorted_indices],
            sparse_min_spanning_tree.data[edges_sorted_indices],
        ),
    ).T
```